src/url-params/url-params.js

```diff
@@ -0,0 +1,333 @@
+ * or even adjust the an Experience's control position.
+ *
+ * The plugin works by using the parameters in the URL to find hotspots with matching tags in
+ * the Experience and sending them click events to trigger any applied actions.
+ *
+ * This plugin is licensed under the MIT license. A copy of this license and
+ * the accompanying source code is available at https://github.com/ceros/ceros-plugins
+ */
+(function() {
+    "use strict";
+
+    /**
+     * Class to manage state represented in the URL
+     *
+     * @constructor
+     */
+    var UrlHashManager = function() {
+
+        var hash = window.location.hash;
+
+        // The State represented in the URL, and the order its keys were added
+        this.state = [];
+
+        // If the current has has anything in it
+        if (hash.length > 0) {
+
+            // Remove leading # symbol
+            hash = hash.replace(/^#/, "");
+
+            // Spit string into key=value pairs
+            var items = hash.split("&");
+
+            // For every pair
+            for (var i = 0; i < items.length; i++){
+
+                // Split into key and value
+                var subItems = items[i].split("=");
+
+                // If there were two
+                if (subItems.length == 2) {
+
+                    this.setParameterValueForKey(
+                        decodeURIComponent(subItems[0]),
+                        decodeURIComponent(subItems[1])
+                    );
+
+                }
+            }
+        }
+    };
+
+    UrlHashManager.prototype = {
+
+        /**
+         * Get Studio tags required to activated to represent current state
+         *
+         * @returns {Array}
+         */
+        getTagsForCurrentState: function(){
+
+            var tags = [];
+
+            // Build array of tags
+            this.state.forEach(function(tag) {
+                tags.push(tag.key + "=" +  tag.value);
+            });
+
+            return tags;
+        },
+
+        /**
+         * Build a string representing the current state for use in the URL
+         *
+         * @returns {string}
+         */
+        buildHashFromCurrentState: function() {
+
+            var value = "",
+                i = 0;
+
+            this.state.forEach(function(tag) {
+
+                // If not first iteration
+                if (i != 0) {
+                    value += "&";
+                }
+
+                // Encode values to make the URL safe
+                value += encodeURIComponent(tag.key) + "=" + encodeURIComponent(tag.value);
+
+                i++;
+            });
+
+            return value;
+        },
+
+        /**
+         * Set item for use in URL
+         *
+         * @param {string} theKey
+         * @param {string} theValue
+         */
+        setParameterValueForKey: function(theKey, theValue) {
+
+            // If we have this key already
+            if (this.hasKey(theKey)) {
```

Comment:
ahh sorry tom, one thing i did notice is that we loop through the state for has key, then loop through again to find the right one.

its probably not a huge deal because the size of state is very small, but i dont know if you need the haskey method. essentially you loop through them all, if the key of the current index is the key you want to set then set it and return, otherwise if you get all the way through the looop without returning you make a new object and push it onto state.

just a little optimization, really not necessary, ill give a +1 without it, but let me know if you want to make this change and i'll re +1